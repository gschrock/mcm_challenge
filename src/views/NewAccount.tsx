import React, { useState } from "react";
import { Button, Col, Form } from "react-bootstrap";
import { useHistory } from "react-router-dom";
import validator from "validator";

interface Props {
  loanResponse: { [key: string]: string | boolean } | null;
}

export const NewAccount: React.SFC<Props> = ({ loanResponse }) => {
  let history = useHistory();
  const [values, setValues] = useState<{
    [key: string]: { value: string; valid: boolean };
  }>({
    username: {
      value: "",
      valid: true
    },
    password: {
      value: "",
      valid: true
    },
    repeat: {
      value: "",
      valid: true
    }
  });

  /**
   * If user attempts to navigate here without loan
   * response data redirect them back to landing route.
   */
  if (!loanResponse) history.push("/");

  const handleInputChange = (e: React.FormEvent<HTMLInputElement>) => {
    const { name, value } = e.currentTarget;
    setValues({
      ...values,
      [name]: {
        value,
        valid: values[name].valid
      }
    });
  };

  const handleSubmit = () => {
    const valid = isFormValid();

    if (valid) {
      alert(
        "We probably want to handle creating an actual account in the future!"
      );
    }
  };

  const isFormValid = () => {
    const regex = /^[A-Za-z]+$/;

    /**
     * Check if passwords contain a number or special character
     * and have desired length
     */
    const validPassword =
      !regex.test(values.password.value) && values.password.value.length >= 8;
    const validRepeatPassword =
      !regex.test(values.repeat.value) &&
      values.repeat.value.length >= 8 &&
      validator.equals(values.repeat.value, values.password.value);
    setValues({
      username: {
        ...values.username,
        valid: validator.isEmail(values.username.value)
      },
      password: {
        ...values.password,
        valid: validPassword
      },
      repeat: {
        ...values.repeat,
        valid: validRepeatPassword
      }
    });
    return (
      validator.isEmail(values.username.value) &&
      validPassword &&
      validRepeatPassword
    );
  };

  return (
    <div className={"new-account-container"}>
      <Form.Group as={Col} controlId="validationCustom01">
        <Form.Label>Username (email)</Form.Label>
        <Form.Control
          onChange={(e: React.FormEvent<HTMLInputElement>) => {
            handleInputChange(e);
          }}
          isInvalid={!values.username.valid}
          type="text"
          placeholder="Username (email)"
          required
        />
        <Form.Control.Feedback type="invalid">
          Please provide a valid email.
        </Form.Control.Feedback>
      </Form.Group>
      <Form.Group as={Col} controlId="validationCustom02">
        <Form.Label>Password</Form.Label>
        <Form.Control
          onChange={(e: React.FormEvent<HTMLInputElement>) => {
            handleInputChange(e);
          }}
          isInvalid={!values.password.valid}
          name={"password"}
          type="password"
          placeholder="Password"
          required
        />
        <Form.Control.Feedback type="invalid">
          Password must contain 8 or more characters and include a number or
          special character.
        </Form.Control.Feedback>
      </Form.Group>
      <Form.Group as={Col} controlId="validationCustom03">
        <Form.Label>Repeat Password</Form.Label>
        <Form.Control
          onChange={(e: React.FormEvent<HTMLInputElement>) => {
            handleInputChange(e);
          }}
          isInvalid={!values.repeat.valid}
          name={"repeat"}
          type="password"
          placeholder="Repeat Password"
          required
        />
        <Form.Control.Feedback type="invalid">
          Passwords must match and contain 8 or more characters and include a
          number or special character.
        </Form.Control.Feedback>
      </Form.Group>
      <div className={"button-container password-button"}>
        <Button onClick={() => handleSubmit()} type="submit">
          Create account
        </Button>
      </div>
    </div>
  );
};
