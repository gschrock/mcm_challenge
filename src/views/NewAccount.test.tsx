import { shallow } from "enzyme";
import React from "react";
import { NewAccount } from "./NewAccount";

jest.mock("react-router-dom", () => ({
  useHistory: () => ({
    push: jest.fn()
  })
}));

describe("NewAccount view", () => {
  it("renders", () => {
    const wrapper = shallow(<NewAccount loanResponse={{ qualified: true }} />);
    expect(wrapper).toMatchSnapshot();
  });
});
