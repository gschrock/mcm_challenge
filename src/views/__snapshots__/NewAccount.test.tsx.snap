// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewAccount view renders 1`] = `
<div
  className="new-account-container"
>
  <FormGroup
    as={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "displayName": "Col",
        "render": [Function],
      }
    }
    controlId="validationCustom01"
  >
    <FormLabel
      column={false}
      srOnly={false}
    >
      Username (email)
    </FormLabel>
    <FormControl
      isInvalid={false}
      name="username"
      onChange={[Function]}
      placeholder="Username (email)"
      required={true}
      type="text"
    />
    <Feedback
      type="invalid"
    >
      Please provide a valid email.
    </Feedback>
  </FormGroup>
  <FormGroup
    as={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "displayName": "Col",
        "render": [Function],
      }
    }
    controlId="validationCustom02"
  >
    <FormLabel
      column={false}
      srOnly={false}
    >
      Password
    </FormLabel>
    <FormControl
      isInvalid={false}
      name="password"
      onChange={[Function]}
      placeholder="Password"
      required={true}
      type="password"
    />
    <Feedback
      type="invalid"
    >
      Password must contain 8 or more characters and include a number or special character.
    </Feedback>
  </FormGroup>
  <FormGroup
    as={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "displayName": "Col",
        "render": [Function],
      }
    }
    controlId="validationCustom03"
  >
    <FormLabel
      column={false}
      srOnly={false}
    >
      Repeat Password
    </FormLabel>
    <FormControl
      isInvalid={false}
      name="repeat"
      onChange={[Function]}
      placeholder="Repeat Password"
      required={true}
      type="password"
    />
    <Feedback
      type="invalid"
    >
      Passwords must match and contain 8 or more characters and include a number or special character.
    </Feedback>
  </FormGroup>
  <div
    className="button-container password-button"
  >
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      type="submit"
      variant="primary"
    >
      Create account
    </Button>
  </div>
</div>
`;
